tests/drivers/i2c/i2c_emul/src/test_forwarding_buf.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "emulated_target.hpp"
+#include <cstdint>
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/fff.h>
+#include <zephyr/ztest.h>
+
+namespace
+{
+
+#define GET_TARGET_DEVICE(node_id, prop, n) DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node_id, prop, n)),
+
+/* Get the devicetree constants */
+constexpr const struct device *controller = DEVICE_DT_GET(CONTROLLER_LABEL);
+constexpr const struct device *targets[FORWARD_COUNT] = {
+	DT_FOREACH_PROP_ELEM(CONTROLLER_LABEL, forwards, GET_TARGET_DEVICE)};
+
+ZTEST(i2c_emul_forwarding, test_write_is_forwarded)
+{
+	uint8_t data[] = {0x00, 0x01, 0x02};
+
+	target_buf_write_received_0_fake.custom_fake = [&data](struct i2c_target_config *,
+							       uint8_t *buf, uint32_t len) {
+		zassert_equal(ARRAY_SIZE(data), len);
+		zexpect_mem_equal(data, buf, len);
+	};
+
+	zassert_ok(
+		i2c_write(controller, data, ARRAY_SIZE(data), emulated_target_config[0].address));
+
+	// Expect 0 reads and 1 write/stop to be made
+	zexpect_equal(0, target_buf_read_requested_0_fake.call_count);
+	zexpect_equal(1, target_buf_write_received_0_fake.call_count);
+	zexpect_equal(1, target_stop_0_fake.call_count);
+}
+
+ZTEST(i2c_emul_forwarding, test_read_is_forwarded)
+{
+	uint8_t expected[] = {0x01, 0x02, 0x03};
+	uint8_t data[ARRAY_SIZE(expected)] = {};
+	target_buf_read_requested_0_fake.custom_fake = [&expected](struct i2c_target_config *,
```

Comment:
I have no idea what ```[&expected](...)``` means here, this needs some commentary

Reply:
Added comments about lambdas & capture variables.